File: Kernel/memory.c
#include <stdint.h>

typedef uint64_t size_t;

void memset(void* ptr, uint8_t value, size_t size) 
{
	for (size_t i = 0; i < size; i++) { ((uint8_t*)ptr)[i] = value; }
}

void memcpy(void* dest, void* src, size_t size) 
{
	for (size_t i = 0; i < size; i++) { ((uint8_t*)dest)[i] = ((uint8_t*)src)[i]; }
}

int8_t memcmp(void* l, void* r, size_t size) 
{
	for (size_t i = 0; i < size; i++) 
	{
		int8_t diff = ((int8_t*)l)[i] - ((int8_t*)r)[i];
		if (diff == 0) { continue; }
		return diff;
	}

	return 0;
}


#define MAGIC 0x69511569
#define NEXTPOWER(x, power) (((x / power) + 1) * power)

typedef struct data 
{
	uint64_t magic;
	size_t size; // excluding sizeof(data)
	// uint8_t pointer[size]; // actual data here
} data;

typedef struct freezone freezone;
struct freezone 
{
	freezone* prevfree;
	freezone* nextfree;
	data* data;
};

freezone base = { .data = 0 };

extern uint8_t KERNEL_END[];
extern void printf(const char* format, ...);

void dumpzones(void) 
{
	printf("--------------------------------------------------------------------------------");

	freezone* zone = &base;

	// get to the left-most zone
	while (zone->prevfree != 0) 
	{
		zone = zone->prevfree;
	}

	while (zone != 0) 
	{
		printf("zone: %x at %x\n", zone->data->size, zone->data);

		zone = zone->nextfree;
	}
}

void error(const char* message, uint64_t arg) 
{
	printf("%s - %x\n", message, arg);
	dumpzones();
	asm("cli");
	asm("hlt");
}

void initialize(void) 
{
	base.nextfree = 0;
	base.prevfree = 0;
	base.data = (data*)KERNEL_END;
	// base.data = (data*)(((uint64_t)KERNEL_END % 16 != 0) ? (void*)NEXTPOWER((uint64_t)KERNEL_END, 16) : KERNEL_END);
	// if ((uint64_t)base.data % 16 != 0) 
	// {
	// 	base.data = (data*)NEXTPOWER((uint64_t)KERNEL_END, 16);
	// }

	base.data->magic = MAGIC;
	base.data->size = (4096 * 100) - sizeof(data);
}

void* kmalloc(size_t size) 
{
	if (base.data == 0) { initialize(); }

	freezone* zone = &base;

	// if (size % 16 != 0) 
	// {
	// 	size = NEXTPOWER(size, 16);
	// }

	// get to the left-most zone
	while (zone->prevfree != 0) 
	{
		zone = zone->prevfree;
	}

	size_t blockSize = sizeof(data) + size;

	// get to first zone with enough size (going from left-most to right-most)
	while (zone != 0) 
	{
		if (zone->data->size == size) 
		{
			if (zone->data->magic != MAGIC) { error("kmalloc: invalid magic", (uint64_t)zone->data->magic); }

			zone->nextfree->prevfree = zone->prevfree;
			zone->prevfree->nextfree = zone->nextfree;

			return zone->data + 1;
		}

		if (zone->data->size > blockSize) 
		{
			if (zone->data->magic != MAGIC) { error("kmalloc: invalid magic", (uint64_t)zone->data->magic); }

			uint8_t* end = (uint8_t*)(zone->data) + sizeof(data) + zone->data->size;
			end -= blockSize;

			((data*)end)->magic = MAGIC;
			((data*)end)->size = size;

			zone->data->size -= blockSize;

			return end + sizeof(data);
		}

		zone = zone->nextfree;
	}

	error("kmalloc: out of memory", (uint64_t)size);
	return 0;
}

void kfree(void* pointer) 
{
	if (base.data == 0) { initialize(); }

	data* datapointer = (data*)pointer - 1;

	if (datapointer->magic != MAGIC) { error("kfree: invalid magic", (uint64_t)datapointer->magic); }

	freezone* zone = &base;

	// get to the right-most zone
	while (zone->nextfree != 0) 
	{
		zone = zone->nextfree;
	}

	freezone* newzone = (freezone*)kmalloc(sizeof(freezone));

	zone->nextfree = newzone;
	newzone->nextfree = 0;
	newzone->prevfree = zone;
	newzone->data = datapointer;
}
















// #include <stdint.h>

// typedef uint64_t size_t;

// void memset(void* ptr, uint8_t value, size_t size) 
// {
// 	for (size_t i = 0; i < size; i++) { ((uint8_t*)ptr)[i] = value; }
// }

// void memcpy(void* dest, void* src, size_t size) 
// {
// 	for (size_t i = 0; i < size; i++) { ((uint8_t*)dest)[i] = ((uint8_t*)src)[i]; }
// }

// int8_t memcmp(void* l, void* r, size_t size) 
// {
// 	for (size_t i = 0; i < size; i++) 
// 	{
// 		int8_t diff = ((int8_t*)l)[i] - ((int8_t*)r)[i];
// 		if (diff == 0) { continue; }
// 		return diff;
// 	}

// 	return 0;
// }

// // https://wiki.osdev.org/Memory_Allocation#A_very_very_simple_Memory_Manager

// typedef struct freezone freezone;
// struct freezone 
// {
// 	freezone* prevfree;
// 	freezone* nextfree;
// 	void* ptr;
// 	size_t size;
// };

// extern uint8_t KERNEL_END[];
// freezone base = { .prevfree = 0, .nextfree = 0, .ptr = 0, .size = 0 };

// void memory_initialize(void) 
// {
// 	base.prevfree = 0;
// 	base.nextfree = 0;
// 	base.ptr = KERNEL_END;
// 	base.size = 1000 * 4096;
// }

// void* kmalloc(size_t size) 
// {
// 	if (base.ptr == 0) { memory_initialize(); }

// 	freezone* zone = &base;

// 	// get to the first freezone
// 	for (; zone->prevfree != 0; zone = zone->prevfree) 
// 	{
// 		if (zone->prevfree == 0) { break; }
// 	}

// 	// get first freezone which has enough size
// 	for (; zone->size < size + sizeof(size_t); zone = zone->nextfree) 
// 	{
// 		if (zone->nextfree == 0) 
// 		{
// 			// "kmalloc(%u): out of memory", size
// 			asm("int3");
// 		}
// 	}

// 	*((size_t*)(zone->ptr)) = size;

// 	void* ptr = zone->ptr + sizeof(size_t);
// 	zone->ptr += size + sizeof(size_t);
// 	zone->size -= size + sizeof(size_t);
// 	return ptr;
// }

// void kfree(void* ptr) 
// {
// 	if (base.ptr == 0) { memory_initialize(); }

// 	freezone* newzone = (freezone*)kmalloc(sizeof(freezone));
// 	newzone->ptr = ptr - sizeof(size_t);
// 	newzone->size = *((size_t*)(ptr - sizeof(size_t))) + sizeof(size_t);

// 	freezone* zone = &base;

// 	// get to the first freezone
// 	for (; zone->prevfree != 0; zone = zone->prevfree) 
// 	{
// 		if (zone->prevfree == 0) { break; }
// 	}

// 	// attach to the bottom of the list
// 	newzone->prevfree = 0;
// 	newzone->nextfree = zone;
// 	zone->prevfree = newzone;
// }

File: Kernel/Memory.cs
global using size_t = uint;

using System;
using System.Runtime;
using System.Runtime.InteropServices;
using System.Runtime.CompilerServices;

namespace Kernel;

public static unsafe class Memory 
{
	private static Freezone basezone;

	[Import] private static extern nint getkernelend();

	private static void Initialize() 
	{
		basezone.NextFree = default;
		basezone.PrevFree = default;
		basezone.Data = (Data*)getkernelend();
		basezone.Data->Magic = Data.MAGIC;
		basezone.Data->Size = (100 * 1024 * 1024) - (size_t)sizeof(Data);
	}

	[Export("kmalloc")]
	public static nint Allocate(size_t size) 
	{
		if (basezone.Data == default)
			Initialize();

		Freezone* zone = (Freezone*)Unsafe.AsPointer<Freezone>(ref basezone);

		// get to the left-most zone
		while (zone->PrevFree != default)
			zone = zone->PrevFree;

		size_t blockSize = (size_t)sizeof(Data) + size;

		// get to first zone with enough size (going from left-most to right-most)
		while (zone != default) 
		{
			if (zone->Data->Size == size) 
			{
				if (zone->Data->Magic != Data.MAGIC)
					throw new InvalidOperationException("Invalid magic.");

				zone->NextFree->PrevFree = zone->PrevFree;
				zone->PrevFree->NextFree = zone->NextFree;

				return (nint)(zone->Data + 1);
			}

			// if (zone->Data->Size > blockSize) 
			// {
			// 	if (zone->Data->Magic != Data.MAGIC)
			// 		throw new InvalidOperationException("Invalid magic.");

			// 	size_t originalSize = zone->Data->Size;

			// 	Data* datapointer = zone->Data;
			// 	datapointer->Size = size;

			// 	zone->Data = (Data*)((byte*)zone->Data + blockSize);
			// 	zone->Data->Magic = Data.MAGIC;
			// 	zone->Data->Size = originalSize - blockSize;

			// 	return (nint)(datapointer + 1);
			// }

			if (zone->Data->Size > blockSize) 
			{
				if (zone->Data->Magic != Data.MAGIC)
					throw new InvalidOperationException("Invalid magic.");

				byte* end = (byte*)(zone->Data) + sizeof(Data) + zone->Data->Size;
				end -= blockSize;

				((Data*)end)->Magic = Data.MAGIC;
				((Data*)end)->Size = size;

				zone->Data->Size -= blockSize;

				return (nint)(end + sizeof(Data));
			}

			zone = zone->NextFree;
		}

		throw new OutOfMemoryException();
	}

	[Export("kfree")]
	public static void Free(nint pointer) 
	{
		if (basezone.Data == default)
			Initialize();

		Data* datapointer = (Data*)pointer - 1;

		if (datapointer->Magic != Data.MAGIC)
			throw new InvalidOperationException("Invalid magic.");

		Freezone* zone = (Freezone*)Unsafe.AsPointer<Freezone>(ref basezone);

		// get to the right-most zone
		while (zone->NextFree != default)
			zone = zone->NextFree;

		Freezone* newzone = (Freezone*)Allocate((size_t)sizeof(Freezone));

		zone->NextFree = newzone;
		newzone->NextFree = default;
		newzone->PrevFree = zone;
		newzone->Data = datapointer;
	}

	[Import(EntryPoint = "memset")] public static extern void Set(nint pointer, byte value, nuint size);
	public static void Set(nint pointer, byte value, size_t size) => Set(pointer, value, (nuint)size);
	public static void Zero(nint pointer, size_t size) => Set(pointer, 0, size);

	[StructLayout(LayoutKind.Sequential, Pack = 4)]
	private struct Data 
	{
		public uint Magic;
		public size_t Size;
		// public fixed byte Pointer[Size]; // actual data here (after Magic and Size, at offset == sizeof(Data))

		public const uint MAGIC = 0x69511569;
	}

	[StructLayout(LayoutKind.Sequential)]
	private struct Freezone 
	{
		public Freezone* PrevFree;
		public Freezone* NextFree;
		public Data* Data;
	}
}
